### PR TITLE
Refactor bootstrap process to use `fluxcd/pkg/git`

### PR DESCRIPTION
**Pull request by aryan9600**
Refactors the bootstrap process to use https://github.com/fluxcd/pkg/tree/main/git/gogit, getting rid of the 
`pkg/bootstrap/git` in the process.

Fixes: #3273
Related to: #3282

Signed-off-by: Sanskar Jaiswal <jaiswalsanskar078@gmail.com>

### CODE REVIEW

File: pkg/bootstrap/bootstrap_plain_git.go [hunk 1]
```diff
@@ -260,60 +282,66 @@ func (b *PlainGitBootstrapper) ReconcileSyncConfig(ctx context.Context, options
...
-
 	// Create secure Kustomize FS
-	fs, err := filesys.MakeFsOnDiskSecureBuild(b.git.Path())
+	fs, err := filesys.MakeFsOnDiskSecureBuild(b.gitClient.Path())
 	if err != nil {
 		return fmt.Errorf("failed to initialize Kustomize file system: %w", err)
 	}
 
 	// Generate Kustomization
 	kusManifests, err := kustomization.Generate(kustomization.Options{
```
Comment (pjbgf) [inline pkg/bootstrap/bootstrap_plain_git.go:292]: As `err = b.git.Write` has been removed, when `kustomization.Generate` gets called the `gotk-sync.yaml` file is not in disk, resulting on incorrect `kustomization.yaml`:

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- gotk-components.yaml
```

**Comment by aryan9600** [comment]
commands and providers tested:

- [x] `flux bootstrap gitub`
- [x] `flux bootstrap gitlab`
- [x] `flux bootstrap git` with GitHub SSH
- [x] `flux bootstrap git` with GitHub HTTPS
- [x] `flux bootstrap git` with GitLab SSH
- [x] `flux bootstrap git` with GitLab HTTPS
- [x] `flux bootstap git` with BitBucket Cloud SSH
- [x] `flux bootstrap git` with BitBucket Cloud HTTPS
- [x] `flux bootstrap git` with AWS CodeCommit SSH
- [x] `flux bootstrap git` with AWS CodeCommit HTTPS
- [x] `flux bootstrap git` with Azure Devops SSH
- [x] `flux bootstrap git` with Azure Devops HTTPS
- [ ] `flux bootstrap git` with BitBucket Server SSH
- [ ] `flux bootstrap git` with BitBucket Server HTTPS

File: go.mod [hunk 2]
```diff
@@ -2,6 +2,12 @@ module github.com/fluxcd/flux2
 
 go 1.18
 
+// Use Flux's version of go-git, which allows clones with multi_ack and multi_ack_detailed.
```
Comment (pjbgf) [inline go.mod:5]: Now that `fluxcd/pkg` migrated to `fluxcd/go-git`, we should update that here to. That can be a commit in itself.

File: pkg/bootstrap/bootstrap_plain_git.go [hunk 3]
```diff
@@ -390,3 +422,31 @@ func (b *PlainGitBootstrapper) ReportComponentsHealth(ctx context.Context, insta
...
+
+func getOpenPgpEntity(keyRing openpgp.EntityList, passphrase, keyID string) (*openpgp.Entity, error) {
+	if len(keyRing) == 0 {
+		return nil, fmt.Errorf("empty GPG key ring")
+	}
+
+	var entity *openpgp.Entity
+	if keyID != "" {
+		for _, ent := range keyRing {
+			if ent.PrimaryKey.KeyIdString() == keyID {
```
Comment (pjbgf) [inline pkg/bootstrap/bootstrap_plain_git.go:434]: This should be case insensitive, we also should trim the `0x` prefix from keyID.

File: pkg/bootstrap/bootstrap_plain_git.go [hunk 4]
```diff
@@ -390,3 +422,31 @@ func (b *PlainGitBootstrapper) ReportComponentsHealth(ctx context.Context, insta
...
+		}
+
+		if entity == nil {
+			return nil, fmt.Errorf("no GPG private key matching key id '%s' found", keyID)
+		}
+	} else {
+		entity = keyRing[0]
+	}
+
+	err := entity.PrivateKey.Decrypt([]byte(passphrase))
```
Comment (pjbgf) [inline pkg/bootstrap/bootstrap_plain_git.go:446]: We must check `entity.PrivateKey` for `nil`, as the keyring may not contain the private key.

**Comment by pjbgf** [comment]
Tested the bootstrap with GPG signing, which worked fine in GitHub:
image

File: pkg/bootstrap/bootstrap_plain_git.go [hunk 5]
```diff
@@ -390,3 +422,39 @@ func (b *PlainGitBootstrapper) ReportComponentsHealth(ctx context.Context, insta
...
+
+		for _, ent := range keyRing {
+			if ent.PrimaryKey.KeyIdString() == keyID {
+				entity = ent
+			}
+		}
+
+		if entity == nil || entity.PrivateKey == nil {
+			return nil, fmt.Errorf("no GPG private key matching key id '%s' found", keyID)
+		}
```
Comment (pjbgf) [inline pkg/bootstrap/bootstrap_plain_git.go:449]: For better debugging, I think this should be two errors:
```suggestion
		if entity == nil {
			return nil, fmt.Errorf("no GPG key matching key id '%s' found", keyID)
		}
		if entity.PrivateKey == nil {
			return nil, fmt.Errorf("keyring does not contain private key for key id '%s'", keyID)
		}
```

**Review by pjbgf — APPROVED** [review]
LGTM

Great stuff @aryan9600! :rocket:

**Review by stefanprodan — APPROVED** [review]
LGTM

Thanks @aryan9600 🏅

File: pkg/bootstrap/bootstrap_plain_git.go [hunk 6]
```diff
@@ -276,6 +288,19 @@ func (b *PlainGitBootstrapper) ReconcileSyncConfig(ctx context.Context, options
 		return err
 	}
 
+	sshUsername, err := b.promptSSHUsername(ctx)
```
Comment (pjbgf) [inline pkg/bootstrap/bootstrap_plain_git.go:291]: Could do with some documentation as to why this has been done and what providers require this.

**Comment by pjbgf** [comment]
@stefanprodan please review the latest changes from an UX PoV. They were required due to the CodeCommit workflow for non `--token-auth`.

**Comment by stefanprodan** [comment]
@aryan9600 @pjbgf I don't understand why is the SSH username a prompt instead of a flag to set it. I also don't understand how and why is this used. Please post an example for AWS CodeCommit to show case what this does.

File: cmd/flux/bootstrap_git.go [hunk 7]
```diff
@@ -62,6 +60,9 @@ command will perform an upgrade if needed.`,
 
   # Run bootstrap for a Git repository with a private key and password
   flux bootstrap git --url=ssh://git@example.com/repository.git --private-key-file=<path/to/private.key> --password=<password>
+
+  # Run bootstrap for a Git repository on AWS CodeCommit
+  flux bootstrap git --url=ssh://<SSH-Key-ID>@git-codecommit.<region>.amazonaws.com/v1/repos/repository --private-key-file=<path/to/private.key> --password=<password>
```
Comment (stefanprodan) [inline cmd/flux/bootstrap_git.go:65]:
```suggestion
  flux bootstrap git --url=ssh://<SSH-Key-ID>@git-codecommit.<region>.amazonaws.com/v1/repos/<repository> --private-key-file=<path/to/private.key> --password=<SSH-passphrase>
```

File: cmd/flux/bootstrap_git.go [hunk 8]
```diff
@@ -145,7 +159,28 @@ func bootstrapGitCmdRun(cmd *cobra.Command, args []string) error {
...
+			return fmt.Errorf("unable to read TLS CA file: %w", err)
+		}
+	}
+	authOpts, err := getAuthOpts(repositoryURL, caBundle)
+	if err != nil {
+		return fmt.Errorf("failed to create authentication options for %s: %w", repositoryURL.String(), err)
+	}
+
+	clientOpts := []gogit.ClientOption{gogit.WithDiskStorage()}
+	if authOpts.Transport == git.HTTP {
```
Comment (pjbgf) [inline cmd/flux/bootstrap_git.go:177]: I think this should be guarded by the flag we already have in place:
```suggestion
	if gitArgs.insecureHttpAllowed {
```

File: cmd/flux/bootstrap_git.go [hunk 9]
```diff
@@ -116,9 +117,22 @@ func bootstrapGitCmdRun(cmd *cobra.Command, args []string) error {
 	if err != nil {
 		return err
 	}
-	gitAuth, err := transportForURL(repositoryURL)
-	if err != nil {
-		return err
+
+	if strings.Contains(repositoryURL.Hostname(), "git-codecommit") && strings.Contains(repositoryURL.Hostname(), "amazonaws.com") {
+		if repositoryURL.Scheme == "ssh" {
```
Comment (pjbgf) [inline cmd/flux/bootstrap_git.go:122]: nit: instead of using literals, we could use the schemes defined in the git package: `git.{SSH|HTTP|HTTPS}`.

**Review by pjbgf — APPROVED** [review]
LGTM

Awesome work @aryan9600! :bow:

File: cmd/flux/bootstrap_git.go [hunk 10]
```diff
@@ -62,6 +60,9 @@ command will perform an upgrade if needed.`,
 
   # Run bootstrap for a Git repository with a private key and password
   flux bootstrap git --url=ssh://git@example.com/repository.git --private-key-file=<path/to/private.key> --password=<password>
+
+  # Run bootstrap for a Git repository on AWS CodeCommit
+  flux bootstrap git --url=ssh://<SSH-Key-ID>@git-codecommit.<region>.amazonaws.com/v1/repos/<repository> --private-key-file=<path/to/private.key> --password=<SSH-passphrase>
```
Comment (stefanprodan) [inline cmd/flux/bootstrap_git.go:65]: Let's add here also an example for Azure DevOps please.

**Review by stefanprodan — APPROVED** [review]
LGTM

Great work 🥇

**Pull request merged**